File: 프로그래머스/level2/게임맵최단거리.py
from collections import deque
def solution(maps):
    
    # distance = [[0]*len(maps[0]) for _ in range(len(maps))] 
    
    
    n = len(maps)
    m = len(maps[0])
    distance = [[0 for _ in range(m)]  for _ in range(n)]
	
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    
    ## 처음 시작값을 큐에 삽입
    
    find = deque([(0, 0)]) 
    maps[0][0] = 0
    distance[0][0] = 1 ## 시작점은 1로
    
    
    while find:
        x,y = find.popleft()
        
        ## 네 방향 확인
        for i in range(4):
            X = x + dx[i]
            Y = y + dy[i]
            ### 범위안에 있으면서 벽이 아닌경우 ##
            if 0 <= X < len(maps[0]) and 0 <= Y < len(maps) and maps[Y][X] == 1:
                maps[Y][X] = 0 ## 중복을 막기 위해 현재위치 벽으로 지정
                distance[Y][X] = distance[y][x] + 1 ## 이전 위치 +1 
                find.append((X,Y))
        
        
    if distance[n-1][m-1] == 0: ## 목적지에 도달하지 못함
        return -1
    else :
        return distance[n-1][m-1]

File: 프로그래머스/level2/게임맵최단거리.java
import java.util.*;

class Solution {
    static void init(int[][] maps){
        for(int i=0; i<maps.length; i++){
            for(int j=0; j<maps[0].length; j++){
                grid[i][j] = maps[i][j];
                distance[i][j] = 0;
            }
        }
    }
    static void myprint(int n, int m){
        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                System.out.print(distance[i][j] + " ");
            }
            System.out.println();
        }
    }
    static class Point{
        int x;
        int y;
        
        public Point(int x, int y){
            this.x = x;
            this.y = y;
        }
    }
    static int[][] grid;
    static int[][] distance;
    static int dx[] = {0, 0, -1, 1};
    static int dy[] = {-1, 1, 0, 0};
    static boolean isRange(int x, int y, int N, int M){
        if(x < 0 || M<=x)
            return false;
        if(y < 0 || N<=y)
            return false;
        return true;
    }
    static void bfs(Point start){
        Queue<Point>q = new LinkedList<>();
        q.add(start);
        distance[start.y][start.x] = 1;
        
        while(!q.isEmpty()){
            Point p = q.poll();
            for(int i=0; i<4; i++){
                int nx = p.x + dx[i];
                int ny = p.y + dy[i];
                
                if(isRange(nx, ny, grid.length, grid[0].length) == true 
                        && distance[ny][nx] == 0 && grid[ny][nx] == 1){
                    distance[ny][nx] = distance[p.y][p.x]+1; 
                    q.add(new Point(nx, ny));
                }
            }
        }
    }
    public int solution(int[][] maps) {
        grid = new int[maps.length][maps[0].length];
        distance = new int[maps.length][maps[0].length];
        
        init(maps);
        bfs(new Point(0, 0));
        int answer = distance[maps.length-1][maps[0].length-1];
        
        if(answer == 0)
            answer= -1;
        
        return answer;
    }
}